server/server.go

```diff
@@ -69,36 +74,74 @@ func (s *Server) prep() {
 //
 // Specifying one of CertFile/KeyFile without the other will panic.
 func (s *Server) ListenAndServe() error {
+	if s.HandleOSSignals {
+		return s.listenAndServeHandleOSSignals()
+	}
+
+	return s.listenAndServe()
+}
+
+// ListenAndServeTLS sets KeyFile and CertFile, then calls ListenAndServe
+func (s *Server) ListenAndServeTLS(certFile, keyFile string) error {
+	if len(certFile) == 0 || len(keyFile) == 0 {
+		panic("either CertFile/KeyFile must be blank, or both provided")
+	}
+	s.KeyFile = keyFile
+	s.CertFile = certFile
+	return s.ListenAndServe()
+}
+
+// Shutdown will gracefully shutdown the server, using a default shutdown
+// timeout if a context is not provided.
+func (s *Server) Shutdown(ctx context.Context) error {
+
+	if ctx == nil {
+		ctx, _ = context.WithTimeout(context.Background(), s.DefaultShutdownTimeout)
+	}
+
+	return s.Server.Shutdown(ctx)
+}
+
+func (s *Server) listenAndServe() error {
+
 	s.prep()
+	if len(s.CertFile) > 0 || len(s.KeyFile) > 0 {
+		return s.Server.ListenAndServeTLS(s.CertFile, s.KeyFile)
+	}
+
+	return s.Server.ListenAndServe()
+}
+
+func (s *Server) listenAndServeHandleOSSignals() error {
 	stop := make(chan os.Signal, 1)
 	signal.Notify(stop, os.Interrupt, os.Kill)
 
+	s.prep()
 	if len(s.CertFile) > 0 || len(s.KeyFile) > 0 {
```

Comment:
could just refactor this bit to move it into `listenAndServeAsync`, and get rid of `listenAndServeAsyncTLS`?

Reply:
Good point. I have made the change and it looks much cleaner.